File: backend/DreamShelf.API/DreamShelf.API/Entities/Book.cs
using DreamShelf.API.Enums;

namespace DreamShelf.API.Entities;

public class Book
{
    public int Id { get; set; }
    public string Title { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;

    public string? CoverImageUrl { get; set; }
    public int? PublicationYear { get; set; }
    public BookStatus BookStatus { get; set; }
    public bool IsPublished { get; set; } = false;
    public AgeRestriction AgeRestriction { get; set; } = AgeRestriction.AllAges;
    public DateTime AddedAt { get; set; } = DateTime.UtcNow;
    public DateTime LastUpdatedAt { get; set; } = DateTime.UtcNow;
    public Role AddedBy { get; set; }
    
    public ICollection<Author> Authors { get; set; } = [];
    public ICollection<Genre> Genres { get; set; } = [];
    public ICollection<Chapter> Chapters { get; set; } = [];
    
    public ICollection<BookImage> BookImages { get; set; } = [];
    
}
// dotnet ef database update
// dotnet ef migrations add UpdateProperties

File: frontend/app/types/book.ts
export interface Author {
  id: number
  displayName: string
}

export interface Genre {
  id: number
  name: string
}

export interface Chapter {
  uuid: string
  title: string
  chapterNumber: number
  lastUpdatedAt: string
}

export interface Book {
  id: number
  title: string
  description: string
  coverImageUrl?: string
  publicationYear?: number
  bookStatus: number
  isPublished: boolean
  ageRestriction: number
  lastUpdatedAt: string
  authors: Author[]
  genres: Genre[]
  chapters: Chapter[]
}

export interface BookStats {
  totalChapters: number
}

export interface UpdateBookData {
  title: string
  description: string
  coverImageUrl: string
  bookStatus: number
  ageRestriction: number
  isPublished: boolean
  genres: number[]
  authorIds: number[]
}

export interface FormBookData {
  title: string
  description: string
  coverImageUrl: string | null
  bookStatus: string
  ageRestriction: number | null
  genres: number[]
  documentFile: File | null
  isPublished: boolean
  documentHeading: string
  authorIds: Author[]
}
